Keen/Dataset/Datasets.cs

```diff
@@ -0,0 +1,301 @@
+                }
+
+                allDefinitions.AddRange(nextSet.Datasets);
+
+            } while (firstSet.Count > allDefinitions.Count);
+
+            return allDefinitions;
+        }
+
+        public async Task DeleteDataset(string datasetName)
+        {
+            if (string.IsNullOrWhiteSpace(datasetName))
+            {
+                throw new KeenException("A dataset name is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var responseMessage = await _keenHttpClient
+                .DeleteAsync(this.GetDatasetUrl(datasetName), _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (responseMessage.StatusCode == HttpStatusCode.NoContent)
+                return;
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+        }
+
+        public async Task<DatasetDefinition> CreateDataset(DatasetDefinition dataset)
+        {
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            //Validate
+            if (dataset == null)
+            {
+                throw new KeenException("An instance of DatasetDefinition must be provided");
+            }
+
+            //This throws if dataset is not valid.
+            dataset.Validate();
+
+            var content = JsonConvert.SerializeObject(dataset, GetSerializerSettings());
+
+            var responseMessage = await _keenHttpClient
+                .PutAsync(this.GetDatasetUrl(dataset.DatasetName), _masterKey, content)
+                .ConfigureAwait(false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinition>(responseString, GetSerializerSettings());
+        }
+
+        /* It's important that there is only one instance of JsonSerializerSettings
+         * for each Serialization/Desierialization to keep the DatasetDefinitionConverter and QueryDefinitionConverter instances threadsafe */
```

Comment:
This keeps those `_isNested` demarcated portions effectively thread-safe I think, if properly ensured, but it seems like it's easier to just let Json.net handle it. I would rather set the converters on the type/property in question where possible assuming we eliminate the complexity. That said, that may not be the most efficient in terms of performance, depending on how json.net creates/destroys those, so let's look into that.